utils/llm.py

```diff
@@ -67,81 +47,68 @@ def ensure_context_len(
 ) -> tuple[LLMContext, int]:
     model = model_from_alias(model)
     max_len = max_len or get_max_prompt_size(model)
-    messages = list()
-    if len(context) > 0 and context[0]["role"] == "system":
+    if context[0]["role"] == "system":
         sys_prompt = context[:1]
         context = context[1:]
     else:
         sys_prompt = []
+        context = context[:]
 
-    context_tokens = count_tokens_for_model(model=model, context=sys_prompt + context)
-    if model.startswith("claude"):
-        context_tokens *= claude_adjust_factor
-
-    reversed_context = list(reversed(context))
-
-    # Latest user's message is always added (there should always be one)
-    messages.append(reversed_context.pop(0))
-
-    # Take messages as pairs and reverse them for the check
-    for message_pair in zip(reversed_context[::2], reversed_context[1::2]):
-        message_tokens = count_tokens_for_model(model=model, context=list(reversed(message_pair)))
-        if model.startswith("claude"):
-            message_tokens *= claude_adjust_factor
-        if context_tokens + message_tokens + response_len > max_len:
+    while True:
+        num_tokens = response_len + count_tokens_for_model(model=model, context=sys_prompt + context)
+        if len(context) <= 1 or num_tokens <= max_len:
             break
-        messages.extend(message_pair)
-        context_tokens += message_tokens
-    messages.reverse()
-    context = sys_prompt + messages
-    # assert len(context) > 1, f"There are messages missing in the context:\n\n{context}"
-    return context, context_tokens
```

Comment:
All this was in here because Claude has some strong ideas about what the context should look like (i.e) no consecutive user, or assistant messages, and a user message has to come first. Maybe give it a test with Claude first?

Reply:
Ah right! I sensed something like that, but yagni posessed me. I think I will just make sure that a user message comes first, and leave the more nuanced handling to the upper layers.

Reply:
Anthropic does not complain anymore about the first non-user message. I had to update haiku's alias though, since they updated it and deprecated the old one.